service/src/main/java/bio/terra/drshub/services/AuthService.java

```diff
@@ -0,0 +1,243 @@
+              }
+              return Stream.of(primaryAuth);
+            })
+        .toList();
+  }
+
+  // Map a GA4GH Authorization type to the types DRSHub keeps in its config.
+  private DrsHubAuthorization mapAccessMethodConfigAuthType(
+      AccessUrlAuthEnum authType,
+      DrsProvider drsProvider,
+      UriComponents components,
+      BearerToken bearerToken) {
+    return switch (authType) {
+      case current_request -> new DrsHubAuthorization(
+          Authorizations.SupportedTypesEnum.BEARERAUTH,
+          accessType -> Optional.ofNullable(bearerToken.getToken()).map(List::of));
+      case fence_token -> new DrsHubAuthorization(
+          Authorizations.SupportedTypesEnum.BEARERAUTH,
+          accessType -> getFenceAccessToken(components.toUriString(), drsProvider, bearerToken));
+      case passport -> new DrsHubAuthorization(
+          Authorizations.SupportedTypesEnum.PASSPORTAUTH,
+          accessType -> fetchPassports(bearerToken));
+    };
+  }
+
+  /**
+   * Reach out to the Drs Provider's options endpoint for the object to get Authorizations required.
+   *
+   * @param drsProvider Drs Provider to reach out to
+   * @param uriComponents DRS URI of the object to be resolved.
+   * @return If the Drs Provider returns an Authorizations, return that in an Optional. Else, return
+   *     the empty Optional.
+   */
+  @VisibleForTesting
+  Optional<Authorizations> fetchDrsAuthorizations(
+      DrsProvider drsProvider, UriComponents uriComponents) {
+    var drsApi = drsApiFactory.getApiFromUriComponents(uriComponents, drsProvider);
+    var objectId = uriComponents.getPath();
+    try {
+      return Optional.ofNullable(drsApi.optionsObject(objectId));
+    } catch (RestClientException ex) {
+      log.warn(
+          "Failed to get authorizations for {} from OPTIONS endpoint for DRS Provider {}. "
+              + "Falling back to configured authorizations",
+          objectId,
+          drsProvider.getName());
+      return Optional.empty();
+    }
+  }
+
+  // Reach out to Bond and get the fence token for the user.
+  private Optional<List<String>> getFenceAccessToken(
+      String drsUri, DrsProvider drsProvider, BearerToken bearerToken) {
+    log.info(
+        "Requesting Bond access token for '{}' from '{}'",
+        drsUri,
+        drsProvider.getBondProvider().orElseThrow());
+
+    var bondApi = bondApiFactory.getApi(bearerToken);
+
+    var response =
+        bondApi.getLinkAccessToken(drsProvider.getBondProvider().orElseThrow().getUriValue());
+
+    return Optional.ofNullable(response.getToken()).map(List::of);
+  }
+
+  /**
+   * Reach out to ECM and get the RAS Passport for the current user
+   *
+   * @param bearerToken token for the current user
+   * @return An Optional list of passports tied to the user.
+   */
+  public Optional<List<String>> fetchPassports(BearerToken bearerToken) {
+    return passportCache.computeIfAbsent(
+        bearerToken.getToken(),
+        token -> {
```

Comment:
this is cool! 

so the bearerToken get reused between requests?

also, would this be more accurate? (I'm not sure myself)
```suggestion
    return passportCache.computeIfAbsent(
        bearerToken.getToken(),
        passport -> {
```

Reply:
So, the arg here is still a token. Line 225 is the key to use to get an existing entry. `token` on line 226 is the token from the execution of `bearer.getToken()` if there's no matching entry, and is used as the key in the new entry created by the anonymous function.

Reply:
oh! that makes sense now!